keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,338 @@
+   * [Updating the discovery document](#updating-the-discovery-document)
+   * [CRDs](#crds)
+   * [Aggregated API servers](#aggregated-api-servers)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Risks](#risks)
+* [Graduation Plan](#graduation-plan)
+* [FAQ](#faq)
+* [Future work: solving other coordination problems](#future-work-solving-other-coordination-problems)
+* [Alternatives](#alternatives)
+   * [Exposing if API servers in an HA setup are using the same storage version](#exposing-if-api-servers-in-an-ha-setup-are-using-the-same-storage-version)
+   * [Letting the storage migrator detect if API server instances are in agreement](#letting-the-storage-migrator-detect-if-api-server-instances-are-in-agreement)
+* [References](#references)
+
+## Overview
+
+During the rolling update of an HA master, API server instances *i)* use
+different storage versions for a built-in API resource, and *ii)* show different
+storageVersionHash in the discovery documents. These facts make the storage
+migrator not working properly ([details][]).
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+We propose a mechanism for HA API servers to vote on the storage version to use,
+then show the selected storage version in the discovery document and switch to
+use the selected version atomically.
+
+## API changes
+
+### Storage Version API
+
+We introduce a new API `StorageVersion`, in a new API group
+`kube-apiserver.internal.k8s.io/v1alpha1`. 
+
+```golang
+// Storage version for a specific resource.
+type StorageVersion struct {
+  TypeMeta
+  // The name is the GVR, i.e., <group>.<version>.<resource>.
+  ObjectMeta
+  // The selected storage version. All API server intances encode the resource
+  // objects in this version when committing it to etcd.
+  SelectedVersion string
+  // Proposed storage version candidate, keyed by the ID of the proposing API
+  // server.
+  CandidateVersions []VersionCandidate
+}
+
+type VersionCandidate struct {
+  // The ID of the reporting API server. 
+  APIServerID string
+  // The preferred storage version of the reporting API server. The preferred
+  // storage version is hardcoded in each Kubernetes release.
+  PreferredVersion string
+}
+```
+
+We will discuss how the `StorageVersion` is updated and used in the [Changes to API
+servers](#changes-to-api-servers) section.
+
+## Changes to API servers
+
+### Curating a list of participating API servers in HA master
+
+Currently, API servers maintain such a list in the "kubernetes" endpoints. See
+this [design doc][] for details.
+
+We will formalize the API in another KEP. For the purpose of this KEP, all we
+need to know is that HA master is able to maintain a list of participating API
```

Comment:
ah, neat.  This is how you handle pruning.  Combined with visibility into apiservices, this is sufficient to selectively prune apiservers.